Source/Soot.cpp

```diff
@@ -0,0 +1,179 @@
+}
+
+void
+PeleC::fill_soot_source(
+  amrex::Real time,
+  amrex::Real dt,
+  amrex::MultiFab& state,
+  amrex::MultiFab& soot_src,
+  int ng)
+{
+  BL_PROFILE("PeleC::fill_soot_source()");
+
+  auto const& fact =
+    dynamic_cast<amrex::EBFArrayBoxFactory const&>(state.Factory());
+  auto const& flags = fact.getMultiEBCellFlagFab();
+
+#ifdef _OPENMP
+#pragma omp parallel
+#endif
+  for (amrex::MFIter mfi(soot_src, amrex::TilingIfNotGPU()); mfi.isValid();
+       ++mfi) {
+    const amrex::Box& bx = mfi.growntilebox(ng);
+    const auto& flag_fab = flags[mfi];
+    amrex::FabType typ = flag_fab.getType(bx);
+    if (typ == amrex::FabType::covered) {
+      continue;
+    }
+    amrex::FArrayBox& Sfab = state[mfi];
+    amrex::FArrayBox& soot_fab = soot_src[mfi];
+    auto const& s_arr = Sfab.array();
+    const int nqaux = NQAUX > 0 ? NQAUX : 1;
+    amrex::FArrayBox mu_cc(bx, 1, amrex::The_Async_Arena());
+    amrex::FArrayBox q(bx, QVAR, amrex::The_Async_Arena());
+    amrex::FArrayBox qaux(bx, nqaux, amrex::The_Async_Arena());
+    auto const& q_arr = q.array();
+    auto const& qaux_arr = qaux.array();
+    auto const& mu_arr = mu_cc.array();
+
+    // Get primitives, Q, including (Y, T, p, rho) from conserved state
+    // required for D term
+    {
+      BL_PROFILE("PeleC::ctoprim()");
+      amrex::ParallelFor(
+        bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+          pc_ctoprim(i, j, k, s_arr, q_arr, qaux_arr);
+        });
+    }
+
+    // Compute transport coefficients, coincident with Q
+    {
+      auto const& qar_yin = q.array(QFS);
+      auto const& qar_Tin = q.array(QTEMP);
+      auto const& qar_rhoin = q.array(QRHO);
+      bool get_xi = false;
+      bool get_mu = true;
+      bool get_lam = false;
+      bool get_diag = false;
+      BL_PROFILE("PeleC::get_transport_coeffs()");
+      // Get Transport coefs on GPU.
+      auto const* ltransparm = trans_parms.device_trans_parm();
+      amrex::ParallelFor(
+        bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+          auto trans = pele::physics::PhysicsType::transport();
+          amrex::Real T = qar_Tin(i, j, k);
+          amrex::Real rho = qar_rhoin(i, j, k);
+          amrex::GpuArray<amrex::Real, NUM_SPECIES> Y = {{0.0}};
+          for (int n = 0; n < NUM_SPECIES; ++n) {
+            Y[n] = qar_yin(i, j, k, n);
+          }
+          amrex::Real* diag = nullptr;
+          amrex::Real mu = 0.;
+          amrex::Real xi, lam;
+          trans.transport(
+            get_xi, get_mu, get_lam, get_diag, T, rho, Y.data(), diag, mu, xi,
+            lam, ltransparm);
+          mu_arr(i, j, k) = mu;
+        });
+    }
+    auto const& soot_arr = soot_fab.array();
+    soot_model->computeSootSourceTerm(bx, q_arr, mu_arr, soot_arr, time, dt);
+  }
+}
+
+void
+PeleC::clipSootMoments(amrex::MultiFab& S_new, const int ng)
+{
+  for (amrex::MFIter mfi(S_new, amrex::TilingIfNotGPU()); mfi.isValid();
```

Comment:
same comment here about using the new parallelfor paradigm. It works better on GPU because it allows you to fuse the boxes.